pkg/ring/kv/memberlist/memberlist_client.go

```diff
@@ -432,70 +435,96 @@ func (m *Client) CAS(ctx context.Context, key string, f func(in interface{}) (ou
 		return fmt.Errorf("key too long: %d", len(key))
 	}
 
-	sleep := false
-	for retries := 10; retries > 0; retries-- {
+	var lastError error = nil
+
+outer:
+	for retries := maxCasRetries; retries > 0; retries-- {
 		m.casAttempts.Inc()
 
-		if sleep {
+		if lastError == errNoChangeDetected {
 			// We only get here, if 'f' reports some change, but Merge function reports no change. This can happen
 			// with Ring's merge function, which depends on timestamps (and not the tokens) with 1-second resolution.
 			// By waiting for one second, we hope that Merge will be able to detect change from 'f' function.
 
 			select {
-			case <-time.After(1 * time.Second):
+			case <-time.After(noChangeDetectedRetrySleep):
 				// ok
 			case <-ctx.Done():
-				break
+				lastError = ctx.Err()
+				break outer
 			}
 		}
 
-		val, ver, err := m.get(key)
-		if err != nil {
-			return err
-		}
-
-		out, retry, err := f(val)
+		change, newver, retry, err := m.trySingleCas(key, f)
 		if err != nil {
-			return err
-		}
-
-		if out == nil {
-			// no change to be done
-			return nil
-		}
-
-		// Don't even try
-		r, ok := out.(Mergeable)
-		if !ok {
-			return fmt.Errorf("invalid type: %T, expected Mergeable", out)
-		}
+			level.Debug(util.Logger).Log("msg", "CAS attempt failed", "err", err, "retry", retry)
 
-		// To support detection of removed items from value, we will only allow CAS operation to
-		// succeed if version hasn't changed, i.e. state hasn't changed since running 'f'.
-		change, newver, err := m.mergeValueForKey(key, r, ver)
-		if err != nil && err != errVersionMismatch {
-			return err
-		}
-
-		if newver == 0 {
-			// 'f' didn't do any update that we can merge. Let's give it a new chance?
+			lastError = err
 			if !retry {
 				break
 			}
-
-			// If merge failed because of version mismatch, don't sleep.
-			sleep = (err != errVersionMismatch)
 			continue
```

Comment:
Collision happens when update to value is received via gossip during CAS call. Those gossips can be received at any time and are not strictly related to timing of changes, so sleeping will not help in that situation. Fortunately, since CAS operates only locally and is generally fast, collisions are rare.

The sleep is there for a different reason -- to make CAS be able to detect the change. This is done via Merge function of the ring, which uses Unix timestamps with one second resolution to detect changes. That's why we wait for one second, if no change is detected.